### PR TITLE
feat: free devices for autosell bot

### PR DESCRIPTION
**Pull request by antmat**
@3Hren @sshaman1101 @oopcode PTAL
I doubt that this API looks ok , so this is done for existing bot compatibility. IMO we should rework this in future.

### CODE REVIEW

File: insonmnia/hardware/hardware.go [hunk 1]
```diff
@@ -146,56 +147,141 @@ func (m *benchValue) Value() uint64 {
...
 	}
-
-	result := bench.GetResult()
+	target := to[id]
 	switch bench.SplittingAlgorithm {
 	case sonm.SplittingAlgorithm_NONE:
-		if to[id].IsSet() {
-			return nil, fmt.Errorf("duplicate benchmark with id %d and type none", bench.ID)
+		if wasSet {
+			panic(fmt.Errorf("duplicate benchmark with id %d and type none", bench.ID))
```
Comment (oopcode) [inline insonmnia/hardware/hardware.go]: why exactly do we panic here (instead of returning an error)?

File: insonmnia/hardware/hardware.go [hunk 2]
```diff
@@ -146,56 +147,141 @@ func (m *benchValue) Value() uint64 {
...
+func (h *Hardware) ResourcesToBenchmarkMap(resources *sonm.AskPlanResources) (benchmarks map[uint64]*sonm.Benchmark, err error) {
 	if !resources.GPU.Normalized() {
 		return nil, errors.New("passed resources are not normalized, call resources.GPU.Normalize(hardware) first")
 	}
-	var err error
-	benchmarks := make([]benchValue, sonm.MinNumBenchmarks)
 
-	proportions := []float64{}
-	hwBenches := []map[uint64]*sonm.Benchmark{}
+	defer func() {
```
Comment (oopcode) [inline insonmnia/hardware/hardware.go]: is this exception handling that we deserved?

Comment (antmat) [reply insonmnia/hardware/hardware.go]: yup

File: insonmnia/hardware/hardware.go [hunk 3]
```diff
@@ -146,56 +147,141 @@ func (m *benchValue) Value() uint64 {
...
+			return nil, fmt.Errorf("unknown hash in passed resources: %s", hash)
+		}
+	}
+
+	proportion := float64(resources.GetCPU().GetCorePercents()) / float64(h.CPU.GetDevice().GetCores()) / 100
+	for _, bench := range h.CPU.GetBenchmarks() {
+		insertBench(hardware.CPU.Benchmarks, bench, proportion)
+	}
 
+	proportion = float64(resources.GetStorage().GetSize().GetBytes()) / float64(h.Storage.GetDevice().GetBytesAvailable())
```
Comment (sshaman1101) [inline insonmnia/hardware/hardware.go]: seems like division by zero may happen

Comment (oopcode) [reply insonmnia/hardware/hardware.go]: it's checked for Nan and Inf, no worries

**Comment by antmat** [comment]
img

**Pull request merged**